libcloud/compute/drivers/oneandone.py

```diff
@@ -0,0 +1,2263 @@
+        host and protocol components.
+        """
+        action = self.api_prefix + action.lstrip('/')
+
+        return super(OneAndOneConnection, self). \
+            request(action=action,
+                    params=params,
+                    data=data,
+                    headers=headers,
+                    method=method,
+                    raw=raw)
+
+
+class OneAndOneNodeDriver(NodeDriver):
+    """
+      Base OneAndOne node driver.
+    """
+    connectionCls = OneAndOneConnection
+    name = '1and1'
+    website = 'http://www.1and1.com'
+    type = Provider.ONEANDONE
+
+    NODE_STATE_MAP = {
+        'POWERING_ON': NodeState.STARTING,
+        'POWERING_OFF': NodeState.PENDING,
+        'POWERED_OFF': NodeState.STOPPING,
+        'POWERED_ON': NodeState.RUNNING,
+        'REBOOTING': NodeState.REBOOTING,
+        'CONFIGURING': NodeState.RECONFIGURING,
+        'REMOVING': NodeState.UNKNOWN,
+        'DEPLOYING': NodeState.STARTING,
+    }
+
+    """
+    Core Functions
+    """
+
+    def list_sizes(self):
+        """
+        Lists all sizes
+
+        :return: A list of all configurable node sizes.
+        :rtype: ``list`` of :class:`NodeSize`
+        """
+        sizes = []
+
+        fixed_instances = self._list_fixed_instances()
+        for value in fixed_instances:
+            node_size = self._to_node_size(value)
+            sizes.append(node_size)
+
+        return sizes
+
+    def list_locations(self):
+        """
+        Lists all locations
+
+        :return: ``list`` of :class:`NodeLocation`
+        :rtype: ``list``
+        """
+        datacenters = self.ex_list_datacenters()
+        locations = []
+        for values in datacenters:
+            node_size = self._to_location(values)
+            locations.append(node_size)
+
+        return locations
+
+    def list_images(self, image_type=None):
+        """
+        :return: ``list`` of :class: `NodeImage`
+        :rtype: ``list``
+        """
+        response = self.connection.request(
+            action='server_appliances',
+            method='GET'
+        )
+
+        return self._to_images(response.object, image_type)
+
+    def get_image(self, image_id):
+        response = self.connection.request(
+            action='server_appliances/%s' % image_id,
+            method='GET'
+        )
+        return self._to_image(response.object)
+
+    """
+    Node functions
+    """
+
+    def create_node(self,
+                    name,
+                    image,
+                    ex_fixed_instance_size_id,
+                    location=None,
+                    auth=None,
+                    ex_ip=None,
+                    ex_monitoring_policy_id=None,
+                    ex_firewall_policy_id=None,
+                    ex_loadbalancer_id=None,
+                    ex_description=None,
+                    ex_power_on=None):
+        """
+        Creates a node.
+
+        :param name: The name of the new node
+        :type ```str```
```

Comment:
like this - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/dimensiondata.py#L894-L916